nighthawk/source/client/client.cc

```diff
@@ -0,0 +1,275 @@
+    int worker_number = 0;
+    while (workers_.size() < concurrency) {
+      const auto worker_delay = std::chrono::duration_cast<std::chrono::nanoseconds>(
+          ((inter_worker_delay_usec * worker_number) * 1us));
+      workers_.push_back(std::make_unique<ClientWorkerImpl>(
+          api_, tls_, benchmark_client_factory_, sequencer_factory_, uri, store_factory_.create(),
+          worker_number, first_worker_start + worker_delay));
+      worker_number++;
+    }
+    return workers_;
+  }
+
+  bool runWorkers() {
+    Envoy::Runtime::RandomGeneratorImpl generator;
+    Envoy::Runtime::ScopedLoaderSingleton loader(
+        Envoy::Runtime::LoaderPtr{new Envoy::Runtime::LoaderImpl(generator, store(), tls())});
+    for (auto& w : workers_) {
+      w->start();
+    }
+    bool ok = true;
+    for (auto& w : workers_) {
+      w->waitForCompletion();
+      ok = ok && w->success();
+    }
+    return ok;
+  }
+
+private:
+  Envoy::Thread::ThreadFactoryImplPosix thread_factory_;
+  Envoy::Filesystem::InstanceImplPosix file_system_;
+  Envoy::Event::RealTimeSystem time_system_;
+  StoreFactoryImpl store_factory_;
+  Envoy::Stats::StorePtr store_;
+  Envoy::Api::Impl api_;
+  Envoy::ThreadLocal::InstanceImpl tls_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  std::vector<ClientWorkerPtr> workers_;
+  const Envoy::Cleanup cleanup_;
+  const BenchmarkClientFactoryImpl benchmark_client_factory_;
+  const SequencerFactoryImpl sequencer_factory_;
+  const Options& options_;
+};
+
+bool Main::runWorkers(ProcessContext& context, std::vector<StatisticPtr>& merged_statistics,
+                      std::map<std::string, uint64_t>& merged_counters) const {
+  Uri uri = Uri::Parse(options_->uri());
+  try {
+    // TODO(oschaaf): DnsLookupFamily should be optionized.
+    uri.resolve(context.dispatcher(), Envoy::Network::DnsLookupFamily::Auto);
+  } catch (UriException) {
+    return false;
+  }
+  const std::vector<ClientWorkerPtr>& workers = context.createWorkers(uri, determineConcurrency());
+  if (context.runWorkers()) {
+    StatisticFactoryImpl statistic_factory(*options_);
+    merged_statistics = mergeWorkerStatistics(statistic_factory, workers);
+    merged_counters = mergeWorkerCounters(workers);
+    return true;
+  }
+  return false;
+}
+
+void Main::writeOutput(ProcessContext& context, const std::vector<StatisticPtr>& merged_statistics,
+                       const std::map<std::string, uint64_t>& merged_counters) const {
+  ConsoleOutputFormatterImpl console_formatter(context.time_system(), *options_, merged_statistics,
+                                               merged_counters);
+  // TODO(oschaaf): output format, location, method, etc should be optionized.
+  std::cout << console_formatter.toString();
+  JsonOutputFormatterImpl json_formatter(context.time_system(), *options_, merged_statistics,
+                                         merged_counters);
+  mkdir("measurements", 0777);
```

Comment:
As with tests below, when doing syscalls, can you `RELEASE_ASSERT` on them (or check and handle errors, seems like this should be what we do in this one, but fine to leave as assert and TODO).